x/gov/keeper.go

```diff
@@ -0,0 +1,415 @@
+	return proposalID
+}
+
+func (keeper Keeper) activateVotingPeriod(ctx sdk.Context, proposal *Proposal) {
+	proposal.VotingStartBlock = ctx.BlockHeight()
+	proposal.Status = "VotingPeriod"
+	keeper.SetProposal(ctx, proposal)
+	keeper.ActiveProposalQueuePush(ctx, proposal)
+}
+
+// =====================================================
+// Procedures
+
+// Gets procedure from store. TODO: move to global param store and allow for updating of this
+func (keeper Keeper) GetDepositProcedure(ctx sdk.Context) *DepositProcedure {
+	return &DepositProcedure{
+		MinDeposit:       sdk.Coins{{"steak", 10}},
+		MaxDepositPeriod: 200,
+	}
+}
+
+// Gets procedure from store. TODO: move to global param store and allow for updating of this
+func (keeper Keeper) GetVotingProcedure(ctx sdk.Context) *VotingProcedure {
+	return &VotingProcedure{
+		VotingPeriod: 200,
+	}
+}
+
+// Gets procedure from store. TODO: move to global param store and allow for updating of this
+func (keeper Keeper) GetTallyingProcedure(ctx sdk.Context) *TallyingProcedure {
+	return &TallyingProcedure{
+		Threshold:         sdk.NewRat(1, 2),
+		Veto:              sdk.NewRat(1, 3),
+		GovernancePenalty: sdk.NewRat(1, 100),
+	}
+}
+
+// =====================================================
+// Votes
+
+// Gets the vote of a specific voter on a specific proposal
+func (keeper Keeper) AddVote(ctx sdk.Context, proposalID int64, voter sdk.Address, option string) sdk.Error {
+	proposal := keeper.GetProposal(ctx, proposalID)
+	if proposal == nil {
+		return ErrUnknownProposal(proposalID)
+	}
+	if proposal.Status != "VotingPeriod" {
+		return ErrInactiveProposal(proposalID)
+	}
+
+	if option != "Yes" && option != "Abstain" && option != "No" && option != "NoWithVeto" {
+		return ErrInvalidVote(option)
+	}
+
+	vote := Vote{
+		ProposalID: proposalID,
+		Voter:      voter,
+		Option:     option,
+	}
+	keeper.setVote(ctx, proposalID, voter, vote)
+
+	return nil
+}
+
+// Gets the vote of a specific voter on a specific proposal
+func (keeper Keeper) GetVote(ctx sdk.Context, proposalID int64, voter sdk.Address) *Vote {
+	store := ctx.KVStore(keeper.storeKey)
+	bz := store.Get([]byte(fmt.Sprintf("%d", proposalID) + ":votes:" + fmt.Sprintf("%s", voter)))
+	if bz == nil {
+		return nil
+	}
+	vote := &Vote{}
+	keeper.cdc.MustUnmarshalBinary(bz, vote)
+	return vote
+}
+
+// Gets the vote of a specific voter on a specific proposal
+func (keeper Keeper) setVote(ctx sdk.Context, proposalID int64, voter sdk.Address, vote Vote) {
+	store := ctx.KVStore(keeper.storeKey)
+	bz := keeper.cdc.MustMarshalBinary(vote)
+	key := []byte(fmt.Sprintf("%d", proposalID) + ":votes:" + fmt.Sprintf("%s", voter))
+	store.Set(key, bz)
+}
+
+// Gets the vote of a specific voter on a specific proposal
+func (keeper Keeper) GetVotes(ctx sdk.Context, proposalID int64) sdk.Iterator {
+	store := ctx.KVStore(keeper.storeKey)
+	return sdk.KVStorePrefixIterator(store, []byte(fmt.Sprintf("%d", proposalID)+":votes:"))
```

Comment:
Put the whole string in `fmt.Sprintf`, and this should be in the store keys file I think

Reply:
Oh I totally second this!